File: arrays and hashing/twosum.py
from collections import defaultdict


class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hasher = defaultdict(int)
        hasher[(target-nums[0])] += 0
        i = 1
        while i < len(nums):
            if nums[i] in hasher:
                return[hasher[nums[i]], i]
            else:
                hasher[(target-nums[i])] += i
            i+=1

sol = Solution()
print(sol.twoSum(nums = [0,3,2,7,11,15], target = 9))

File: arrays and hashing/twosum.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;


class Solution{
    public int[] twoSum(int[] numbers, int target) {
        int[] result = new int[2];
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < numbers.length; i++) {
            if (map.containsKey(target - numbers[i])) {
                result[1] = i;
                result[0] = map.get(target - numbers[i]);
                return result;
            }
            map.put(numbers[i], i);
        }
        return result;
    }

    public static void main(String[] args){
        int[] numbers = new int[]{ 2,7,11,15 };
        int target = 9;
        int sol[] = new Solution().twoSum(numbers, target);
        System.out.println(Arrays.toString(sol));
    }
}